trie/sync.go

```diff
@@ -425,6 +446,41 @@ func (s *Sync) children(req *nodeRequest, object node) ([]*nodeRequest, error) {
 			node: node.Val,
 			path: append(append([]byte(nil), req.path...), key...),
 		}}
+		// Mark all internal nodes between shortNode and its **in disk**
+		// child as invalid. This is essential in the case of path mode
+		// scheme; otherwise, state healing might overwrite existing child
+		// nodes silently while leaving a dangling parent node within the
+		// range of this internal path on disk. This would break the
+		// guarantee for state healing.
+		//
+		// While it's possible for this shortNode to overwrite a previously
+		// existing full node, the other branches of the fullNode can be
+		// retained as they remain untouched and complete.
+		//
+		// This step is only necessary for path mode, as there is no deletion
+		// in hash mode at all.
+		if _, ok := node.Val.(hashNode); ok && s.scheme == rawdb.PathScheme {
+			owner, inner := ResolvePath(req.path)
+			for i := 1; i < len(key); i++ {
+				// Theoretically, it's necessary to check for the presence before
+				// blindly caching deletion commands. However, due to the fact that
+				// Pebble doesn't use a bloom filter to enhance read performance
+				// for non-existent items, this check would significantly slow down
+				// overall performance. FIX IT(rjl493456442)
+				var exists bool
+				if owner == (common.Hash{}) {
+					exists = rawdb.ExistsAccountTrieNode(s.database, append(inner, key[:i]...))
+				} else {
+					exists = rawdb.ExistsStorageTrieNode(s.database, owner, append(inner, key[:i]...))
+				}
+				if exists {
+					req.deletes = append(req.deletes, key[:i])
```

Comment:
This tripped me up. In `req.deletes`, you store `key[:i]`, which is the key which this shortnode (extension-node) is holding. However, the actual full key is larger, and you use `inner` to construct the full one: 
```
rawdb.ExistsAccountTrieNode(s.database, append(inner, key[:i]...))
```

So, a bit later in this code, when going from `req.deletes -> membatch.deletes`, you convert the "partial paths" into full paths by prefixing:

```
for _, segment := range req.deletes {
		path := append(req.path, segment...)
		s.membatch.deletes[string(path)] = struct{}{}
		s.membatch.size += uint64(len(path))
	}
```

I missed the `req.deletes -> membatch.deletes` conversion the first time around, and couldn't figure out how on earth the snippet which does the deletion could work, using non-full paths:
```
	for path := range s.membatch.deletes {
		owner, inner := ResolvePath([]byte(path))
		rawdb.DeleteTrieNode(dbw, owner, inner, common.Hash{} /* unused */, s.scheme)
	}
```

I guess you save a bit of memory by not storing the expanded path here, and recalculating it later. So I can't really object, just noting that it makes it a bit more complex.